src/sources/vector/v2.rs

```diff
@@ -42,24 +43,28 @@ impl proto::Service for Service {
                     (event, None)
                 }
             })
-            .ok_or_else(|| Status::invalid_argument("missing event"))?;
-
-        self.pipeline
-            .clone()
-            .send(event)
-            .await
-            .map_err(|err| Status::unavailable(err.to_string()))?;
-
-        let status = if let Some(receiver) = receiver {
-            receiver.await
-        } else {
-            BatchStatus::Delivered
-        };
-        match status {
-            BatchStatus::Delivered => Ok(Response::new(proto::EventResponse {})),
-            BatchStatus::Errored => Err(Status::internal("Delivery error")),
-            BatchStatus::Failed => Err(Status::data_loss("Delivery failed")),
+            .collect();
+
+        for (event, receiver) in items {
+            self.pipeline
+                .clone()
+                .send(event)
+                .await
+                .map_err(|err| Status::unavailable(err.to_string()))?;
+
+            let status = match receiver {
+                Some(receiver) => receiver.await,
+                None => BatchStatus::Delivered,
+            };
+
+            match status {
+                BatchStatus::Errored => return Err(Status::internal("Delivery error")),
+                BatchStatus::Failed => return Err(Status::data_loss("Delivery failed")),
+                BatchStatus::Delivered => (),
+            }
```

Comment:
@bruceg what do you propose should happen here if messages 1 and 2 or 5 are delivered, but 3 of 5 fails to be delivered. Should it stop processing 4 and 5, or try to process all events, and return some kind of status response that identifies which events failed to be delivered?

Reply:
There should be a single `BatchNotifier` created for the whole array of events, which would produce a single status for the entire set. Move creating the notifier (and receiver) up out of the `map` around line 38, and then you should be able to `send_all` the entire batch at once and collect a single status at the end.